feat(Storage): AWSS3PluginPrefixResolver

lawmicha opened the pull request:
_Issue #, if available:_
iOS - https://github.com/aws-amplify/amplify-ios/pull/1277

_Description of changes:_

Add docs for storage prefix resolver

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

royjit commented on docs/lib/storage/fragments/js/configureaccess.md:
````diff
@@ -88,4 +88,4 @@ If you want to have custom *private* path prefix like *myPrivatePrefix/*, you ne
 ```
-This ensures only the authenticated users has the access to the objects under the path.
\ No newline at end of file
+This ensures only the authenticated users has the access to the objects under the path. 
````
Remove the white space at the end.
```suggestion
This ensures only the authenticated users has the access to the objects under the path.
```

royjit commented on docs/lib/storage/fragments/ios/configureaccess/50_customization.md:
```diff
@@ -0,0 +1,136 @@
+import Amplify
+import AmplifyPlugins
+
+struct MyDeveloperDefinedPrefixResolver: AWSS3StoragePluginPrefixResolver {
```
Can we add some comments on what the code is doing? The documentation starts off with a struct `MyDeveloperDefinedPrefixResolver` without giving much info about why we need it, I think it would be better if we explain what resolver are and how we use them. Then show the code?

lawmicha replied:
Will iterate on this, and add some more comments first. I think ihis is actually subsection of a larger doc that has some content about it above it

royjit commented on docs/lib/storage/fragments/ios/configureaccess/50_customization.md:
```diff
@@ -0,0 +1,136 @@
+
+#### Examples
+
+1. See [StorageAccessLevelAwarePrefixResolver](https://github.com/aws-amplify/amplify-ios/blob/main/AmplifyPlugins/Storage/AWSS3StoragePlugin/Configuration/AWSS3PluginPrefixResolver.swift#L26) and [StorageRequestUtils.getAccessLevelPrefix](https://github.com/aws-amplify/amplify-ios/blob/main/AmplifyPlugins/Storage/AWSS3StoragePlugin/Support/Utils/StorageRequestUtils+Getter.swift#L24) for an examp ...
```
Giving code reference like this would break in future when we update the code. Maybe use permalinks?

lawmicha replied:
what do you mean by using a permalink? The feedback I got was to explain the current logic since it would be useful for developers looking into custom resolvers. I think i can address this by explaining it without linking to code snippet

lawmicha replied:
as discussed, will describe the example instead of relying on code snippet here

lawmicha commented:
Rebased from latest, added some comments to the code, made sure the code compiles, removed the file references (they were not useful), cleaned up the use case sections

royjit commented on src/fragments/lib/storage/ios/configureaccess/50_customization.mdx:
```diff
@@ -0,0 +1,142 @@
+
+#### Examples
+
+1. See [StorageAccessLevelAwarePrefixResolver](https://github.com/aws-amplify/amplify-ios/blob/main/AmplifyPlugins/Storage/AWSS3StoragePlugin/Configuration/AWSS3PluginPrefixResolver.swift#L26) and [StorageRequestUtils.getAccessLevelPrefix](https://github.com/aws-amplify/amplify-ios/blob/main/AmplifyPlugins/Storage/AWSS3StoragePlugin/Support/Utils/StorageRequestUtils+Getter.swift#L24) for an examp ...
```
I thought we are removing code snippet examples. Do we really need this point 1, these resolvers are used internally and developers do not need to know about it.

lawmicha replied:
taking a look, might have missed pushing a commit here

lawmicha replied:
ah yes, forgot to push the [commit](https://github.com/aws-amplify/docs/pull/3321/commits/631d945c7624ad8405378ae7f89c8c7b1e0fcb3b) - this removed the code snippet examples

royjit reviewed (approved):
Approving with one comment

lawmicha commented:
@aws-amplify/amplify-js @aws-amplify/amplify-sa can I get a review on this? iOS team has approved already

jakeburden commented on src/fragments/lib/storage/ios/configureaccess/50_customization.mdx:
```diff
@@ -0,0 +1,145 @@
+
+You can also perform validation based on the access controls you have defined. For example, if you have defined Guests with no access then you can fail the request early by checking if the user is not signed in:
+
+<amplify-callout>
```
This is just `<Callout>` now

jakeburden replied:
```suggestion
<Callout>
```

jakeburden commented on src/fragments/lib/storage/ios/configureaccess/50_customization.mdx:
```diff
@@ -0,0 +1,145 @@
+
+<amplify-callout>
+
+This will only stop your app from making a call with an unauthenticated user. You must also set up your IAM policies to protect your resources. See [CLI Storage](~/cli/storage/import.md) for more details. 
```
We have migrated from `md` to `mdx`, this can be linked with `[CLI Storage](/cli/storage/import)`

jakeburden replied:
```suggestion
This will only stop your app from making a call with an unauthenticated user. You must also set up your IAM policies to protect your resources. See [CLI Storage](/cli/storage/import) for more details. 
```

jakeburden commented on src/fragments/lib/storage/ios/configureaccess/50_customization.mdx:
```diff
@@ -0,0 +1,145 @@
+
+This will only stop your app from making a call with an unauthenticated user. You must also set up your IAM policies to protect your resources. See [CLI Storage](~/cli/storage/import.md) for more details. 
+
+</amplify-callout>
```
```suggestion
</Callout>
```

lawmicha commented:
Thanks @jakeburden for the review, rebased and fixed things up

jakeburden commented:
Thanks @lawmicha! I'm creating a staging URL here to visually test it, once it's built (will be around 15 minutes or so)  I'll verify everything looks ready and will approve:
https://feat-storage-prefix.d1oosfztpocl9c.amplifyapp.com/
